### DISCUSSION TITLE
How to use the JavaScript API correctly?

### DISCUSSION THREAD

**Discussion by migueldilaj** [synfron/ReshaperForBurp#21] [Q&A] [answered]
Hi!
Tried a Run Script Then as follows:

<pre>
console.log('Tring to obtain variable');
console.log(getMessageValueKeys());
</pre>

It failed with this error:

<pre>
{
  "message" : "Script Log",
  "payload" : "Tring to obtain variable"
}
{
  "message" : "Failure running rule",
  "exception" : "ReferenceError: \"getMessageValueKeys\" is not defined. (<cmd>#2)\norg.mozilla.javascript.EcmaError: ReferenceError: \"getMessageValueKeys\" is not defined. (<cmd>#2)\n\tat org.mozilla.javascript.ScriptRuntime.constructError(ScriptRuntime.java:4280)\n\tat org.mozilla.javascript.ScriptRuntime.constructError(ScriptRuntime.java:4258)\n\tat org.mozilla.javascript.ScriptRuntime.notFoundError(ScriptRuntime.java:4348)\n\tat org.mozilla.javascript.ScriptRuntime.getNameFunctionAndThis(ScriptRuntime.java:2462)\n\tat org.mozilla.javascript.optimizer.OptRuntime.callName0(OptRuntime.java:85)\n\tat org.mozilla.javascript.gen._cmd__9._c_script_0(<cmd>:2)\n\tat org.mozilla.javascript.gen._cmd__9.call(<cmd>)\n\tat org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:412)\n\tat org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3578)\n\tat org.mozilla.javascript.gen._cmd__9.call(<cmd>)\n\tat org.mozilla.javascript.gen._cmd__9.exec(<cmd>)\n\tat org.mozilla.javascript.Context.evaluateString(Context.java:1234)\n\tat synfron.reshaper.burp.core.rules.thens.ThenRunScript.lambda$perform$0(ThenRunScript.java:25)\n\tat synfron.reshaper.burp.core.rules.thens.entities.script.Dispatcher.lambda$getRunner$0(Dispatcher.java:53)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)\n\tat java.base/java.lang.Thread.run(Thread.java:831)\n",
  "payload" : "\tat <cmd>:2\n"
}
</pre>

Any advise on what I'm doing wrong will be much appreciated!

**Comment by ddwightx** [comment, marked as answer]
Looks you like actually wanted `Reshaper.event.getMessageValueKeys()`. The documentation may be a little confusing in that area, but the section header titles of that page https://synfron.github.io/ReshaperForBurp/ScriptingLibrary.html are the prefixes that need to go before the function call.

**Comment by migueldilaj** [comment]
Perfect! Thanks for the advice. The calls work as intended now.

**Comment by AkikoOrenji** [comment]
Sorry to hijack but i'm struggling with the following 

console.log('Test');
console.log(Reshaper.variables.getGlobalVariable(value4));

Am i missing something as i get the following: 

_{
  "message" : "Failure running rule",
  "exception" : "ReferenceError: \"value4\" is not defined. (<cmd>#2)\norg.mozilla.javascript.EcmaError: ReferenceError: \"value4\" is not defined. (<cmd>#2)\r\n\tat org.mozilla.javascript.ScriptRuntime.constructError(ScriptRuntime.java:4563)\r\n\tat org.mozilla.javascript.ScriptRuntime.constructError(ScriptRuntime.java:4544)\r\n\tat org.mozilla.javascript.ScriptRuntime.notFoundError(ScriptRuntime.java:4632)\r\n\tat org.mozilla.javascript.ScriptRuntime.name(ScriptRuntime.java:1952)\r\n\tat org.mozilla.javascript.gen._cmd__5._c_script_0(<cmd>:2)\r\n\tat org.mozilla.javascript.gen._cmd__5.call(<cmd>)\r\n\tat org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:380)\r\n\tat org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3868)\r\n\tat org.mozilla.javascript.gen._cmd__5.call(<cmd>)\r\n\tat org.mozilla.javascript.gen._cmd__5.exec(<cmd>)\r\n\tat org.mozilla.javascript.Context.evaluateString(Context.java:1137)\r\n\tat synfron.reshaper.burp.core.rules.thens.ThenRunScript.lambda$perform$0(ThenRunScript.java:26)\r\n\tat synfron.reshaper.burp.core.rules.thens.entities.script.Dispatcher.lambda$getRunner$0(Dispatcher.java:53)\r\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\r\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\r\n\tat java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)\r\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)\r\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)\r\n\tat java.base/java.lang.Thread.run(Thread.java:831)\r\n",
  "payload" : "\tat <cmd>:2\r\n"
}_

I also tried global:value4 but get a different error. I assume i dont need 'global:' suffix as its implied in the function name (getGlobalVariable)

thanks

**Reply by ddwightx** [reply]
Hi @AkikoOrenji `console.log(Reshaper.variables.getGlobalVariable(value4));` should be `console.log(Reshaper.variables.getGlobalVariable("value4"));`

**Reply by AkikoOrenji** [reply]
thanks. Rookie mistake. Working fine now. 
Might as well share my full example incase anyone else comes looking for one. Using the Run Script Then to strip all newlines and CRs from a variable before use in another Then step.

var x = Reshaper.variables.getGlobalVariable("value4")
x = x.replace(/[\n\r]+/g, '');
Reshaper.variables.setGlobalVariable("value5", x)

**Reply by ddwightx** [reply]
@AkikoOrenji FYI, you can also just use Set Variable to do what you need:
1. Add a `Set Variable` then
2. Uncheck `Use Message Value`
3. Set `Source Text` = `{{global:value4}}`
4. Check `Use Regex Replace`
5. Set `Regex Pattern` = `[\n\r]+`
6. Set `Destination Variable Name` = `value5`

**Reply by AkikoOrenji** [reply]
oh thanks for the extra tip.
